File: src/main/java/ru/mydesignstudio/database/metadata/extractor/extractors/fk/ForeignKeyExtractor.java
package ru.mydesignstudio.database.metadata.extractor.extractors.fk;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.extractors.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.extractors.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

import java.util.List;

@Component
public class ForeignKeyExtractor implements TableMetadataAppender {
  @StringResource("classpath:sql/extract_foreign_keys.sql")
  private String extractQuery;

  @Autowired
  private ExtractHelper helper;

  @Override
  public void append(TableMetadata metadata, String schemaName, String tableName) {
    List<ForeignKeyModel> foreignKeys = helper.extract(extractQuery, new Object[]{tableName, schemaName}, ForeignKeyModel.class);
    metadata.setForeignKeys(foreignKeys);
  }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/ForeignKeyExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.ForeignKeyModel

@Component
class ForeignKeyExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<ForeignKeyModel> {
        val query = queryProvider.provide("classpath:sql/extract_foreign_keys.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
            ForeignKeyModel(
                    foreignKey = getString(rs, "foreign_key"),
                    constraintType = getString(rs, "constraint_type"),
                    schemaName = getString(rs, "schema_name"),
                    constraintName = getString(rs, "constraint_name"),
                    tableName = getString(rs, "table_name"),
                    columns = getString(rs, "columns"),
                    rOwner = getString(rs, "r_owner"),
                    rConstraintName = getString(rs, "r_constraint_name")
            )
        }
    }
}